### PR TITLE
Telegraf/InfluxDB receiver

### PR DESCRIPTION
**Pull request by jacobmarble**
**Description:** Add `influxdbreceiver`

This change adds a receiver for [InfluxData line protocol](https://docs.influxdata.com/influxdb/v2.0/reference/syntax/line-protocol/), the native wire transfer protocol for [Telegraf](https://www.influxdata.com/time-series-platform/telegraf/) and [InfluxDB](https://www.influxdata.com/products/influxdb/).

The conversion logic lives in [package `influx2otel`](https://github.com/influxdata/influxdb-observability/tree/main/influx2otel), and is also imported by the [proposed Telegraf OpenTelemetry output plugin](https://github.com/influxdata/telegraf/pull/9228). The objective of this strategy is to create a canonical mapping, and an open-source implementation, of OpenTelemetry <-> Telegraf/InfluxDB translation.

**Link to tracking Issue:** #2951

**Testing:** The translation logic is tested in [package `influx2otel`](https://github.com/influxdata/influxdb-observability/tree/main/influx2otel).

**Documentation:** README.md included in new receiver.

### CODE REVIEW

**Comment by jacobmarble** [comment]
@tigrannajaryan this is the follow-on to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2952

**Comment by jacobmarble** [comment]
Rebased and fixed a pointer thing

File: receiver/influxdbreceiver/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# InfluxDB Receiver
```
Comment (tigrannajaryan) [inline receiver/influxdbreceiver/README.md:1]: 👍  for the nice README.

File: receiver/influxdbreceiver/receiver.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+		w.WriteHeader(http.StatusBadRequest)
+		_, _ = fmt.Fprintf(w, "failed to convert protobuf bytes to OTLP object")
+		return
+	}
+	if err = r.nextConsumer.ConsumeMetrics(req.Context(), md); err != nil {
+		if consumererror.IsPermanent(err) {
+			w.WriteHeader(http.StatusBadRequest)
+		} else {
+			w.WriteHeader(http.StatusInternalServerError)
+		}
```
Comment (tigrannajaryan) [inline receiver/influxdbreceiver/receiver.go:194]: 👍

File: receiver/influxdbreceiver/receiver.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+	nr := mux.NewRouter()
+	nr.HandleFunc("/write", r.handleWrite)        // InfluxDB 1.x
+	nr.HandleFunc("/api/v2/write", r.handleWrite) // InfluxDB 2.x
+
+	r.wg.Add(1)
+	r.server = r.httpServerSettings.ToServer(nr)
+	go func() {
+		defer r.wg.Done()
+		if err := r.server.Serve(ln); err != nil {
```
[the remark this answers is not in the document]
Comment (jacobmarble) [reply receiver/influxdbreceiver/receiver.go:83]:
```
// Serve always returns a non-nil error and closes l.
// After Shutdown or Close, the returned error is ErrServerClosed.
```

Fixed.

**Review by tigrannajaryan — APPROVED** [review]

**Comment by neelayu** [comment]
Hey @jacobmarble 
Thanks for putting this receiver in. I am very new to telegraf, but familiar with opentelemetry. 

I deployed an OpenTelemetry collector with influxdb receiver, along with a telegraf agent reading memory metrics and outputting using the HTTP.

Otelcol yaml
```
receivers:
  influxdb:
    endpoint: 0.0.0.0:9888
```

Telegraf conf:
```
[[inputs.mem]]

[[outputs.http]]
    url = "http://localhost:9888/write"
    data_format = "influx"
```

However, the receiver seems to return 400. I think the problem is with the format of data which the agent is sending. 
Any tips?
Thanks

**Comment by jacobmarble** [comment]
Hi @neelayu , I'll try to reproduce this today.

**Comment by jacobmarble** [comment]
@neelayu please tell me more about your environment. I could not reproduce with the latest checkouts of both projects, or with Telegraf v1.19.2, or with opentelemetry-collector-contrib v0.31.0. FWIW, I'm using MacOS 11.4 and Golang 1.16.5.

Using the latest checkout of opentelemetry-collector-contrib, configured:
```yaml
receivers:
  influxdb:
    endpoint: 0.0.0.0:9888

processors:

exporters:
  file:
    path: /dev/stdout

extensions:

service:
  extensions: []
  pipelines:
    metrics:
      receivers: [influxdb]
      processors: []
      exporters: [file]
```

STDOUT:

```text
$ bin/otelcontribcol_darwin_amd64 --config local/otelcol.yaml --log-level INFO
2021-08-12T10:46:34.360-0700    info    service/collector.go:297        Starting otelcontribcol...      {"Version": "v0.31.0-63-g1cc681441", "NumCPU": 16}
2021-08-12T10:46:34.360-0700    info    service/collector.go:236        Loading configuration...
2021-08-12T10:46:34.361-0700    info    service/collector.go:252        Applying configuration...
2021-08-12T10:46:34.361-0700    info    builder/exporters_builder.go:264        Exporter was built.     {"kind": "exporter", "name": "file"}
2021-08-12T10:46:34.361-0700    info    builder/pipelines_builder.go:214        Pipeline was built.     {"pipeline_name": "metrics", "pipeline_datatype": "metrics"}
2021-08-12T10:46:34.361-0700    info    builder/receivers_builder.go:227        Receiver was built.     {"kind": "receiver", "name": "influxdb", "datatype": "metrics"}
2021-08-12T10:46:34.361-0700    info    service/service.go:143  Starting extensions...
2021-08-12T10:46:34.361-0700    info    service/service.go:188  Starting exporters...
2021-08-12T10:46:34.361-0700    info    builder/exporters_builder.go:93 Exporter is starting... {"kind": "exporter", "name": "file"}
2021-08-12T10:46:34.361-0700    info    builder/exporters_builder.go:98 Exporter started.       {"kind": "exporter", "name": "file"}
2021-08-12T10:46:34.361-0700    info    service/service.go:193  Starting processors...
2021-08-12T10:46:34.2021-08-12T10:46:34.361-0700        info    builder/receivers_builder.go:76 Receiver started.       {"kind": "receiver", "name": "influxdb"}
2021-08-12T10:46:34.361-0700    info    service/collector.go:200        Setting up own telemetry...
2021-08-12T10:46:34.368-0700    info    service/telemetry.go:99 Serving Prometheus metrics      {"address": ":8888", "level": 0, "service.instance.id": "ea26dbe9-d922-4926-9d4a-ff32fa7858d7"}
2021-08-12T10:46:34.368-0700    info    service/collector.go:212        Everything is ready. Begin running and processing data.
{"resourceMetrics":[{"resource":{},"instrumentationLibraryMetrics":[{"instrumentationLibrary":{},"metrics":[{"name":"mem_free","gauge":{"dataPoints":[{"labels":[{"key":"host","value":"Orient.local"}],"timeUnixNano":"1628790460000000000","asInt":"6069084160"}]}},{"name":"mem_available_percent","gauge":{"dataPoints":[{"labels":[{"key":"host","value":"Orient.local"}],"timeUnixNano":"1628790460000000000","asDouble":48.020100593566895}]}},{"name":"mem_total","gauge":{"dataPoints":[{"labels":[{"key":"host","value":"Orient.local"}],"timeUnixNano":"1628790460000000000","asInt":"34359738368"}]}},{"name":"mem_active","gauge":{"dataPoints":[{"labels":[{"key":"host","value":"Orient.local"}],"timeUnixNano":"1628790460000000000","asInt":"11763527680"}]}},{"name":"mem_inactive","gauge":{"dataPoints":[{"labels":[{"key":"host","value":"Orient.local"}],"timeUnixNano":"1628790460000000000","asInt":"10430496768"}]}},{"name":"mem_wired","gauge":{"dataPoints":[{"labels":[{"key":"host","value":"Orient.local"}],"timeUnixNano":"1628790460000000000","asInt":"4417286144"}]}},{"name":"mem_available","gauge":{"dataPoints":[{"labels":[{"key":"host","value":"Orient.local"}],"timeUnixNano":"1628790460000000000","asInt":"16499580928"}]}},{"name":"mem_used","gauge":{"dataPoints":[{"labels":[{"key":"host","value":"Orient.local"}],"timeUnixNano":"1628790460000000000","asInt":"17860157440"}]}},{"name":"mem_used_percent","gauge":{"dataPoints":[{"labels":[{"key":"host","value":"Orient.local"}],"timeUnixNano":"1628790460000000000","asDouble":51.979899406433105}]}}]}]}]}
```

Using the latest checkout of telegraf, configured:
```text
[[inputs.mem]]

[[outputs.http]]
    url = "http://localhost:9888/write"
    data_format = "influx"
```

STDOUT:

```text
$ telegraf --config telegraf-simple.conf 
2021-08-12T17:47:31Z I! Starting Telegraf 
2021-08-12T17:47:31Z I! Loaded inputs: mem
2021-08-12T17:47:31Z I! Loaded aggregators: 
2021-08-12T17:47:31Z I! Loaded processors: 
2021-08-12T17:47:31Z I! Loaded outputs: http
2021-08-12T17:47:31Z I! Tags enabled: host=Orient.local
2021-08-12T17:47:31Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"Orient.local", Flush Interval:10s
```

**Comment by neelayu** [comment]
@jacobmarble Thanks for response. 
My setup is Centos 7, Opentelemetry contrib version 0.28.0 and Telegraf running latest master. 
I think the issue present in v0.28 of OtelColContrib. I tried the latest 0.31 and it is working fine.

```
bin/otelcontribcol_linux_amd64.1 --config=config.yml
2021-08-13T09:35:38.247Z	info	service/application.go:262	Starting otelcontribcol...	{"Version": "v0.27.0-236-gb80aa72ba", "NumCPU": 12}
```

```
./telegraf --config telegraf-simple.conf
2021-08-13T09:35:41Z I! Starting Telegraf
2021-08-13T09:35:41Z I! Loaded inputs: mem
2021-08-13T09:35:41Z I! Loaded aggregators:
2021-08-13T09:35:41Z I! Loaded processors:
2021-08-13T09:35:41Z I! Loaded outputs: http
2021-08-13T09:35:41Z I! Tags enabled: host=neelay-upadhyaya
2021-08-13T09:35:41Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"neelay-upadhyaya", Flush Interval:10s
2021-08-13T09:35:51Z E! [agent] Error writing to outputs.http: when writing to [http://localhost:9888/write] received status code: 400
^C2021-08-13T09:35:55Z I! [agent] Hang on, flushing any cached metrics before shutdown
2021-08-13T09:35:55Z E! [agent] Error writing to outputs.http: when writing to [http://localhost:9888/write] received status code: 400
2021-08-13T09:35:55Z I! [agent] Stopping running outputs
```

**Pull request merged**